src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+
+        if smbus_data.read_write == I2C_SMBUS_READ {
+            unsafe {
+                match smbus_data.size {
+                    I2C_SMBUS_BYTE => reqs[0].buf[0] = (*smbus_data.data).byte,
+                    I2C_SMBUS_BYTE_DATA => reqs[1].buf[0] = (*smbus_data.data).byte,
+                    I2C_SMBUS_WORD_DATA => {
+                        reqs[1].buf[0] = ((*smbus_data.data).word & 0xff) as u8;
+                        reqs[1].buf[1] = ((*smbus_data.data).word >> 8) as u8;
+                    }
+
+                    _ => {
+                        println!("Invalid SMBUS command: {}", smbus_data.size);
+                        return Err(Error::new(EINVAL));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// I2C backend and helpers
+const MAX_I2C_VDEV: usize = 1 << 7;
+const I2C_INVALID_ADAPTER: u32 = 0xFFFFFFFF;
+
+pub struct I2cBackend {
+    adapters: Vec<I2cAdapter>,
+    adapter_map: [u32; MAX_I2C_VDEV],
+}
+
+impl I2cBackend {
+    pub fn new(list: &str, dryrun: bool) -> Result<I2cBackend> {
+        let mut adapter_map: [u32; MAX_I2C_VDEV] = [I2C_INVALID_ADAPTER; MAX_I2C_VDEV];
+        let mut adapters: Vec<I2cAdapter> = Vec::new();
+        let devices: Vec<&str> = list.split(',').collect();
+
+        for (i, dev) in devices.iter().enumerate() {
+            let list: Vec<&str> = dev.split(':').collect();
+            let dev = String::from("/dev/i2c-") + list[0];
+            let clients = &list[1..];
+            let mut adapter = I2cAdapter::new(dev, dryrun)?;
+
+            adapter.bus = list[0].parse::<u32>().map_err(|_| Error::new(EINVAL))?;
+            adapter.get_func(dryrun)?;
+
+            for client in clients {
+                let client = client.parse::<usize>().map_err(|_| Error::new(EINVAL))?;
+
+                if client > MAX_I2C_VDEV {
+                    println!("Invalid client address {}", client);
+                    return Err(Error::new(EADDRNOTAVAIL));
+                }
+
+                if adapter_map[client] != I2C_INVALID_ADAPTER {
+                    println!(
+                        "Client address {} is already used by {}",
+                        client, adapters[adapter_map[client] as usize].bus
+                    );
+                    return Err(Error::new(EADDRINUSE));
+                }
+
+                adapter.set_client_addr(client, dryrun)?;
+                adapter_map[client] = i as u32;
+            }
+
+            println!(
+                "Added I2C master with bus id: {:x} for clients: {:?}",
+                adapter.bus, clients
+            );
+
+            adapters.push(adapter);
+        }
+
+        Ok(I2cBackend {
+            adapters,
+            adapter_map,
+        })
+    }
+
+    pub fn transfer(&self, reqs: &mut [I2cReq]) -> Result<()> {
+        let client = reqs[0].addr as usize;
+        let index = self.adapter_map[client];
```

Comment:
Should validate “client < MAX_I2C_VDEV" before accessing adapter_map[client]? Otherwise a malicious vm may crash the daemon.

Reply:
> Should validate “client < MAX_I2C_VDEV" before accessing adapter_map[client]? Otherwise a malicious vm may crash the daemon.

Good point.